nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
+        mels, mel_lens = self.spectrogram_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        mels_pred, *_, pitch = self.spectrogram_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=mel_lens,
+            input_lens=text_lens,
+        )
+
+        mels = mels.cpu()
+        mel_lens = mel_lens.cpu()
+        mels_pred = mels_pred.cpu()
+
+        batch_size, *_ = mels.shape
+
+        targets = []
+        conditions = []
+        lengths = []
+
+        for i in range(batch_size):
+            l = mel_lens[i].item()
+            lengths.append(l)
+
+            targets.append(mels[i, :, :l].T)
+            conditions.append(mels_pred[i, :, :l].T)
+
+        target = nn.utils.rnn.pad_sequence(targets, batch_first=True)
+        condition = nn.utils.rnn.pad_sequence(conditions, batch_first=True)
+        lengths = torch.LongTensor(lengths).unsqueeze(-1)
+
+        target = rearrange(target, "b l c -> b 1 c l")
+        condition = rearrange(condition, "b l c -> b 1 c l")
+
+        return self.move_to_correct_device((target, condition, lengths))
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        target, condition, lengths = self.prepare_batch(batch)
+
+        with torch.no_grad():
+            target = self.normalize_spectrograms(target)
+            condition = self.normalize_spectrograms(condition)
+
+            target = mask(target, lengths)
+            condition = mask(condition, lengths)
+
+        # train discriminator
+        if optimizer_idx == 0:
+            enhanced_spectrograms = self.forward(condition, lengths, mixing=True, normalize=False)
+            fake_logits = self.D(enhanced_spectrograms, condition, lengths)
+
+            real_images = target.requires_grad_()
+            real_logits = self.D(real_images, condition, lengths)
+            d_loss = hinge_loss(real=real_logits, fake=fake_logits)
+            self.log("d_loss", d_loss, prog_bar=True)
+
+            if batch_idx % 4 == 0:
+                gp_loss = gradient_penalty(real_images, real_logits)
+                self.log("d_loss_gp", gp_loss, prog_bar=True)
+                return d_loss + gp_loss
+
+            return d_loss
+
+        # train generator
+        if optimizer_idx == 1:
+            enhanced_spectrograms = self.forward(condition, lengths, mixing=True, normalize=False)
+            fake_logits = self.D(enhanced_spectrograms, condition, lengths)
+            g_loss = gen_hinge_loss(fake_logits)
+            c_loss = 10 * consistency_loss(condition, enhanced_spectrograms, lengths)
```

Comment:
reason behind 10? Hardcoding should be avoided. Make it a class variable with proper name.
loss should be part of class.

Reply:
Good catch, this is supposed to be a parameter

Reply:
Done

Reply:
`"consistency_loss_weight"`